src/third-party/EnsController.ts

```diff
@@ -0,0 +1,121 @@
+import BaseController, { BaseConfig, BaseState } from '../BaseController';
+import { isValidAddress, toChecksumAddress } from 'ethereumjs-util';
+
+/**
+ * @type EnsEntry
+ *
+ * ENS entry representation
+ *
+ * @property chainId - Id of the associated chain
+ * @property name - Nickname associated with this address
+ * @property address - Hex address of a recipient account
+ */
+export interface EnsEntry {
+	chainId: number;
+	name: string;
+	address: string;
+}
+
+/**
+ * @type EnsState
+ *
+ * ENS controller state
+ *
+ * @property ensEntries - Object of ENS entry objects
+ */
+export interface EnsState extends BaseState {
+	ensEntries: { [chainId: string]: { [name: string]: EnsEntry } };
+}
+
+/**
+ * @type EnsEntriesState
+ *
+ * ENS controller state
+ *
+ * @property ensEntries - Object of ENS entry objects
+ */
+interface EnsEntriesState {
+	[chainId: string]: { [name: string]: EnsEntry };
+}
+
+/**
+ * Controller that manages a list ENS names and their resolved addresses
+ */
+export class EnsController extends BaseController<BaseConfig, EnsState> {
+	/**
+	 * Name of this controller used during composition
+	 */
+	name = 'EnsController';
+
+	/**
+	 * Creates an EnsController instance
+	 *
+	 * @param config - Initial options used to configure this controller
+	 * @param state - Initial state to set on this controller
+	 */
+	constructor(config?: Partial<BaseConfig>, state?: Partial<EnsState>) {
+		super(config, state);
+
+		this.defaultState = { ensEntries: {} };
+
+		this.initialize();
+	}
+
+	/**
+	 * Remove all chain Ids and ENS entries from state
+	 */
+	clear() {
+		this.update({ ensEntries: {} });
+	}
+
+	/**
+	 * Remove a contract entry by address
+	 *
+	 * @param chainId - Parent chain of the ENS entry to delete
+	 * @param name - Name of the ENS entry to delete
+	 */
+	delete(chainId: number, name: string) {
```

Comment:
I think I should also be able to delete by address here? (optionally by address or name)
In the AddressBookController I would need to check if the entry IsEns, store the name, delete the entry in addressBook by address, then also delete from the EnsController by the name.

In addressbook deleting by name is a bit trickier because it could be argued that nothing should stop one from having two entries of the same name. In the Ens case, it wouldn't resolve to two addresses.

Reply:
After offline discussions, we decided that deleting address book entries will not affect the ENS Controller, and we do not need a way of deleting by address.

This also avoids problems of deleting ENS entries by address, when names and addresses go out of sync.